runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,350 @@
+  group = augroup,
+  desc = 'Refresh document highlights on cursor movement',
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    if bufstates[bufnr] then
+      debounced_refresh(bufnr)
+    end
+  end,
+})
+api.nvim_create_autocmd('LspNotify', {
+  group = augroup,
+  desc = 'Refresh document highlights on document change or open',
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+    if
+      client.supports_method(ms.textDocument_documentHighlight, { bufnr = bufnr })
+      and (
+        args.data.method == ms.textDocument_didChange
+        or args.data.method == ms.textDocument_didOpen
+      )
+    then
+      debounced_refresh(bufnr)
+    end
+  end,
+})
+
+---@param kind lsp.DocumentHighlightKind
+---@return string
+local function kind_to_hl_group(kind)
+  local base = 'LspReference'
+  if kind == 2 then
+    return string.format('%sRead', base)
+  elseif kind == 3 then
+    return string.format('%sWrite', base)
+  else -- kind == 1 also the default
+    return string.format('%sText', base)
+  end
+end
+
+api.nvim_set_decoration_provider(namespace, {
+  on_win = function(_, _, bufnr, toprow, botrow)
+    local bufstate = bufstates[bufnr]
+    if not bufstate then
+      return
+    end
+
+    if bufstate.version == bufstate.applied then
+      return
+    end
+
+    local enabled = bufstate.enabled
+    if enabled == nil then
+      enabled = globalstate.enabled
+    end
+
+    if not enabled then
+      return
+    end
+
+    local client_hilights = bufstate.client_hilights
+
+    for row = toprow, botrow do
+      api.nvim_buf_clear_namespace(bufnr, namespace, row, row + 1)
+
+      for _, row_highlights in pairs(client_hilights) do
+        local highlights = row_highlights[row] or {}
+
+        for _, highlight in pairs(highlights) do
+          local col = highlight.range['start'].character
+          local end_row = highlight.range['end'].line
+          local end_col = highlight.range['end'].character
+
+          api.nvim_buf_set_extmark(bufnr, namespace, row, col, {
+            hl_group = kind_to_hl_group(highlight.kind),
+            ephemeral = false,
+            end_row = end_row,
+            end_col = end_col,
+          })
+        end
+      end
+    end
+
+    bufstate.applied = bufstate.version
+  end,
+})
+
+---Optional filters |kwargs|, or `nil` for all.
+---@class vim.lsp.document_highlight.enable.Filter
+---@inlinedoc
+---
+---Buffer number, or 0 for current buffer, or nil for all.
+---@field bufnr? integer
+
+---Query whether document highlight is enabled in the {filter}ed scope
+---@param filter? vim.lsp.document_highlight.enable.Filter
+---@return boolean
+function M.is_enabled(filter)
+  vim.validate({ filter = { filter, 'table', true } })
+  filter = filter or {}
+
+  local bufnr = filter.bufnr
+  if bufnr == nil then
+    return globalstate.enabled
+  else
```

Comment:
```suggestion
  end
```

Don't nest after a early return.